### PR TITLE
Optimisation: Add zero-garbage deserialiser for ByteBuffer to RoaringBitmap

### PR DESCRIPTION
**Pull request by shikharid**
### SUMMARY
- There is no direct way to convert a byte array or a ByteBuffer (representing an uncompressed bitmap) to a RoaringBitmap 
- The best way is to create a BitSet first and then use: BitSetUtil.bitmapOf(bitset)
- This creates unnecessary heap garbage as, essentially, a full copy of the byte array is created for BitSet
- This PR introduces a method in BitSetUtil that can be used to directly convert a ByteBuffer to RoaringBitmap
- The implementation tries to do it with constant and minimum possible memory allocation
- This is very useful for performance sensitive code doing this very frequently. Now a RoaringBitmap can be created directly from bytes read on wire with almost no unnecessary memory allocs
- For testing, BitSetUtil tests are replicated and benchmarks are added

### RESULTS
- Benchmark Results are [posted here](https://github.com/RoaringBitmap/RoaringBitmap/pull/650#issuecomment-1667252580)
- **TL;DR** 
- Average time is upto 10-20% faster when we go from small to larger bitsets (compared to existing way)
- GC pressure is about 4-5x lower for all size types

### Automated Checks

- [x] I have run `./gradlew test` and made sure that my PR does not break any unit test.
- [x] I have run `./gradlew checkstyleMain` or the equivalent and corrected the formatting warnings reported.

### CODE REVIEW

**Comment by shikharid** [comment]
# Benchmark Setup

- JMH version: 1.23
- VM version: JDK 1.8.0_342, OpenJDK 64-Bit Server VM, 25.342-b07
- VM invoker: ***/Java/JavaVirtualMachines/liberica-1.8.0_342/jre/bin/java
- OS: MacOS Ventura 13.2.1
- Arch: Apple M1 Max (32 gb, aarch64)
- JMH Threads: 1
- Warmup: 30 sec (6 rounds of 5 sec)
- Measure: 60 sec (6 rounds of 10 sec)
- Forks: 1


# small bitsets, wordSize = 64 represents 4096 bits (512 bytes)

```
Benchmark                                                                        Mode  Cnt           Score           Error   Units
BitSetUtilBenchmark.ByteArrayToBitsetToRoaring                                   avgt    6      931086.885 ?     20283.562   us/op
BitSetUtilBenchmark.ByteArrayToRoaring                                           avgt    6     1185022.469 ?      3664.055   us/op

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.alloc.rate                    avgt    6        2898.212 ?        60.073  MB/sec
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.alloc.rate                            avgt    6         630.085 ?         1.893  MB/sec

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.alloc.rate.norm               avgt    6  2969120079.273 ?         0.001    B/op
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.alloc.rate.norm                       avgt    6   820106024.000 ?         0.001    B/op

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.churn.PS_Eden_Space           avgt    6        2905.461 ?       268.548  MB/sec
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.churn.PS_Eden_Space                   avgt    6         645.967 ?       151.244  MB/sec

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.churn.PS_Eden_Space.norm      avgt    6  2976684836.364 ? 281648933.014    B/op
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.churn.PS_Eden_Space.norm              avgt    6   840783189.333 ? 197060626.323    B/op

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.churn.PS_Survivor_Space       avgt    6           0.035 ?         0.028  MB/sec
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.churn.PS_Survivor_Space               avgt    6           0.011 ?         0.013  MB/sec

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.churn.PS_Survivor_Space.norm  avgt    6       35746.909 ?     29415.418    B/op
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.churn.PS_Survivor_Space.norm          avgt    6       13956.741 ?     16356.828    B/op

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.count                         avgt    6          89.000                  counts
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.count                                 avgt    6          29.000                  counts

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.time                          avgt    6         317.000                      ms
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.time                                  avgt    6         133.000                      ms
```


# midsized bitsets, wordSize = 512 represents 32768 bits (~4kb)

```
Benchmark                                                                        Mode  Cnt           Score           Error   Units
BitSetUtilBenchmark.ByteArrayToBitsetToRoaring                                   avgt    6     1017930.073 ?     12592.306   us/op
BitSetUtilBenchmark.ByteArrayToRoaring                                           avgt    6      904132.343 ?     14992.134   us/op

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.alloc.rate                    avgt    6        2528.769 ?        30.500  MB/sec
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.alloc.rate                            avgt    6         666.502 ?        13.209  MB/sec

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.alloc.rate.norm               avgt    6  2833754699.200 ?         0.001    B/op
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.alloc.rate.norm                       avgt    6   661747621.030 ?         7.077    B/op

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.churn.PS_Eden_Space           avgt    6        2511.728 ?       318.397  MB/sec
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.churn.PS_Eden_Space                   avgt    6         668.619 ?        87.415  MB/sec

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.churn.PS_Eden_Space.norm      avgt    6  2814736247.467 ? 361833709.120    B/op
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.churn.PS_Eden_Space.norm              avgt    6   664027301.495 ?  98479990.345    B/op

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.churn.PS_Survivor_Space       avgt    6           0.029 ?         0.026  MB/sec
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.churn.PS_Survivor_Space               avgt    6           0.012 ?         0.016  MB/sec

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.churn.PS_Survivor_Space.norm  avgt    6       32768.000 ?     29057.408    B/op
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.churn.PS_Survivor_Space.norm          avgt    6       11915.636 ?     15509.537    B/op

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.count                         avgt    6          83.000                  counts
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.count                                 avgt    6          30.000                  counts

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.time                          avgt    6         294.000                      ms
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.time                                  avgt    6         144.000                      ms



# large bitsets, wordSize = 8192 represents 524288 bits (~64kb)

Benchmark                                                                        Mode  Cnt           Score           Error   Units
BitSetUtilBenchmark.ByteArrayToBitsetToRoaring                                   avgt    6      979362.981 ?      9706.318   us/op
BitSetUtilBenchmark.ByteArrayToRoaring                                           avgt    6      848216.153 ?     13435.627   us/op

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.alloc.rate                    avgt    6        2847.291 ?        26.830  MB/sec
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.alloc.rate                            avgt    6         978.851 ?        14.645  MB/sec

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.alloc.rate.norm               avgt    6  3061785079.273 ?         0.001    B/op
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.alloc.rate.norm                       avgt    6   913781812.444 ?         3.053    B/op

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.churn.PS_Eden_Space           avgt    6        2848.020 ?       176.047  MB/sec
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.churn.PS_Eden_Space                   avgt    6         980.787 ?       192.524  MB/sec

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.churn.PS_Eden_Space.norm      avgt    6  3062596576.970 ? 191595150.892    B/op
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.churn.PS_Eden_Space.norm              avgt    6   915610737.778 ? 179449894.470    B/op

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.churn.PS_Survivor_Space       avgt    6           0.038 ?         0.033  MB/sec
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.churn.PS_Survivor_Space               avgt    6           0.032 ?         0.062  MB/sec

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.churn.PS_Survivor_Space.norm  avgt    6       41217.212 ?     35212.641    B/op
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.churn.PS_Survivor_Space.norm          avgt    6       30037.333 ?     58717.057    B/op

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.count                         avgt    6          80.000                  counts
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.count                                 avgt    6          66.000                  counts

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.time                          avgt    6         298.000                      ms
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.time                                  avgt    6         271.000                      ms
```


# very large bitsets, wordSize = 131072 represents 8388608 bits (~8.3 million, ~1mb)

```
Benchmark                                                                        Mode  Cnt           Score           Error   Units
BitSetUtilBenchmark.ByteArrayToBitsetToRoaring                                   avgt    6     1045351.199 ?     21878.484   us/op
BitSetUtilBenchmark.ByteArrayToRoaring                                           avgt    6      936327.563 ?      7578.909   us/op

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.alloc.rate                    avgt    6        2545.367 ?        60.330  MB/sec
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.alloc.rate                            avgt    6         742.882 ?         5.766  MB/sec

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.alloc.rate.norm               avgt    6  2925470527.600 ?  40249330.630    B/op
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.alloc.rate.norm                       avgt    6   765114215.273 ?         0.001    B/op

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.churn.PS_Eden_Space           avgt    6        2561.021 ?       353.990  MB/sec
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.churn.PS_Eden_Space                   avgt    6         760.875 ?       160.004  MB/sec

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.churn.PS_Eden_Space.norm      avgt    6  2943789738.667 ? 424205667.201    B/op
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.churn.PS_Eden_Space.norm              avgt    6   783596078.545 ? 162411623.855    B/op

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.churn.PS_Survivor_Space       avgt    6           0.431 ?         0.461  MB/sec
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.churn.PS_Survivor_Space               avgt    6           0.132 ?         0.157  MB/sec

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.churn.PS_Survivor_Space.norm  avgt    6      495358.667 ?    536171.478    B/op
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.churn.PS_Survivor_Space.norm          avgt    6      136036.848 ?    161777.609    B/op

BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.count                         avgt    6          63.000                  counts
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.count                                 avgt    6          45.000                  counts
sa
BitSetUtilBenchmark.ByteArrayToBitsetToRoaring:?gc.time                          avgt    6         185.000                      ms
BitSetUtilBenchmark.ByteArrayToRoaring:?gc.time                                  avgt    6         134.000                      ms
```

**Comment by shikharid** [comment]
Also, anyone with more experience with the codebase please confirm this. 

This is probably a bug: https://github.com/RoaringBitmap/RoaringBitmap/pull/650/files#diff-608ac1c40d6f95be23548cf97937dfcee083b21634337f6bb57617565c467f05R163

The copy should be (from, to) and not **(from, from + BLOCK_LENGTH)**

**Comment by shikharid** [comment]
Other approach that avoid use of ThreadLocal's and do everything in-place, maybe slightly more complex implementation:
Choose the right container type by looking ahead in buffer
- Create the char[4096] or long[1024] by first scanning through the ByteBuffer and counting upto 4096 set bits.
- Reset bytebuffer to start position
- Create the array/bitmap container and read data from bytebuffer for next 2^16 bits 

We can avoid lookahead by always starting from let's say an ArrayContainer and convert it to a BitmapContainer when the 4096 threshold is crossed. 
Can keep then unused char[] array for next container to optimise allocs. At worst one unused char[4096] array will be declared for dense bitsets

**Comment by lemire** [comment]
Would you consider porting your code over to [‎RoaringBitmap/src/main/java/org/roaringbitmap/buffer/BufferBitSetUtil.java](https://github.com/RoaringBitmap/RoaringBitmap/blob/2669c4f5a49ee7da5ff4cd70e18ee5520018d6a5/RoaringBitmap/src/main/java/org/roaringbitmap/buffer/BufferBitSetUtil.java#L89)

We try to keep them in sync. It should be easy work.

File: RoaringBitmap/src/main/java/org/roaringbitmap/BitSetUtil.java [hunk 1]
```diff
@@ -71,6 +72,71 @@ public static RoaringBitmap bitmapOf(final long[] words) {
     return ans;
   }
 
+  // To avoid memory allocation, reuse ThreadLocal buffers
+  private static final ThreadLocal<long[]> WORD_BLOCK = ThreadLocal.withInitial(() ->
```
Comment (lemire) [inline RoaringBitmap/src/main/java/org/roaringbitmap/BitSetUtil.java:76]: This is 8 kB that may get allocated for each thread for one function (below) and it never gets released.

I don't think we want that.


What would be accessible is to allow the user to (optionally) pass a buffer to the function below. If the user passes a buffer, then you use it, otherwise, you allocate it. 

In this manner, you give the use full control over the performance, and you don't make people pay 8 kB that they don't want to lock down for one function.

File: RoaringBitmap/src/main/java/org/roaringbitmap/BitSetUtil.java [hunk 2]
```diff
@@ -71,6 +72,71 @@ public static RoaringBitmap bitmapOf(final long[] words) {
...
+      blockCardinality += Long.bitCount(word);
+
+      // When block is full, add block to bitmap
+      if (blockLength == BLOCK_LENGTH) {
+        // Each block becomes a single container, if any bit is set
+        if (blockCardinality > 0) {
+          ans.highLowContainer.insertNewKeyValueAt(containerIndex++, Util.highbits(offset),
+              BitSetUtil.containerOf(0, blockLength, blockCardinality, words));
+        }
+        offset += (BLOCK_LENGTH * Long.SIZE);
```
Comment (lemire) [inline RoaringBitmap/src/main/java/org/roaringbitmap/BitSetUtil.java:112]: Though this can be reasonably dismissed, there is the possibility that offset overflows. Make sure that the offset variable cannot overflow (hopefully it cannot due to the the max size of a Java Bitset, but please be specific, maybe with a comment).

**Comment by lemire** [comment]
This is a very good PR. Please consider my comments.

**Comment by shikharid** [comment]
@lemire made the changes you recommended and also ported to BufferBitSetUtil

**Comment by lemire** [comment]
@shikharid 

Regarding...

```Java
 private static Container containerOf(final int from, final int to, final int blockCardinality,
      final long[] words) {
    // find the best container available
    if (blockCardinality <= ArrayContainer.DEFAULT_MAX_SIZE) {
      // containers with DEFAULT_MAX_SIZE or less integers should be
      // ArrayContainers
      return arrayContainerOf(from, to, blockCardinality, words);
    } else {
      // otherwise use bitmap container
      return new BitmapContainer(Arrays.copyOfRange(words, from, from + BLOCK_LENGTH),
          blockCardinality);
    }
  }
```


A possible alternative would be...

```Java
      if(to - from < BLOCK_LENGTH) {
        long [] newbuffer = new long[BLOCK_LENGTH];
        arraycopy(words, from, newbuffer, 0, to - from);
                            Object dest_arr, int destPos, int len)
        return new BitmapContainer(newbuffer, blockCardinality);
      } else {
        return new BitmapContainer(Arrays.copyOfRange(words, from, from + BLOCK_LENGTH),
            blockCardinality);
     }
```

If you'd like to make this change (it will need to be done in both version of containerOf, one in the buffer package and one in the main package), or a related change, that would be fine.

Please advise.

**Comment by shikharid** [comment]
Can't really see if that helps with anything.
OpenJDK 17 implementation of copyOfRange looks like:

```java
public static long[] copyOfRange(long[] original, int from, int to) {
        int newLength = to - from;
        if (newLength < 0)
            throw new IllegalArgumentException(from + " > " + to);
        long[] copy = new long[newLength];
        System.arraycopy(original, from, copy, 0,
                         Math.min(original.length - from, newLength));
        return copy;
    }
```

Doesn't seem to have changed since 8. 

So the changes I made are essentially the same, without the extra if checks:

```java
  private static Container containerOf(final int from, final int to, final int blockCardinality,
      final long[] words) {
    // find the best container available
    if (blockCardinality <= ArrayContainer.DEFAULT_MAX_SIZE) {
      // containers with DEFAULT_MAX_SIZE or less integers should be
      // ArrayContainers
      return arrayContainerOf(from, to, blockCardinality, words);
    } else {
      // otherwise use bitmap container
      long[] container = new long[BLOCK_LENGTH];
      System.arraycopy(words, from, container, 0, to - from);
      return new BitmapContainer(container, blockCardinality);
    }
  }
```

File: RoaringBitmap/src/main/java/org/roaringbitmap/BitSetUtil.java [hunk 3]
```diff
@@ -15,7 +16,7 @@ public class BitSetUtil {
 
   // a block consists has a maximum of 1024 words, each representing 64 bits,
   // thus representing at maximum 65536 bits
-  static final private int BLOCK_LENGTH = BitmapContainer.MAX_CAPACITY / Long.SIZE; //
+  public static final int BLOCK_LENGTH = BitmapContainer.MAX_CAPACITY / Long.SIZE; //
```
Comment (lemire) [inline RoaringBitmap/src/main/java/org/roaringbitmap/BitSetUtil.java:19]: Is it necessary to make BLOCK_LENGTH public?

Comment (shikharid) [reply RoaringBitmap/src/main/java/org/roaringbitmap/BitSetUtil.java:19]: Didn't see any other neat way to expose the information. 
Since the user can provide the buffer, they need to know atleast what size it needs to be. 
Can just mention it in Javadoc, the bounds check will anyways raise error if bad sized buffer is provided.

**Review by lemire — APPROVED** [review]

**Comment by shikharid** [comment]
Hey @lemire, any plans of merging/releasing the soon?
Added it as I wanted this for my usecase, so would be nice if this was in release and I don't have to use a fork.

**Comment by lemire** [comment]
Merging. I will issue a release.
